### DISCUSSION TITLE
Accessibility to allow blind organists to use GrandOrgue

### DISCUSSION THREAD

**Discussion by lewisalexander2020** [GrandOrgue/grandorgue#1553] [General]
Hi there, 

I'm  writing this as a blind organist, on a mac using VoiceOver. I am also a developer behind a speaking digital organ solution and whilst working with a particular software developer I cannot disclose for legal reasons, I wanted to investigate accessibility support for GrandOrgue in MacOS.

Initial setup allows me to navigate menus and setup resources, but the main organ console panel has no degree of support what so ever, meaning that VoiceOver doesn't even see the window nor have any ability to navigate it, interact with any elements, etc.

I'd like to help where I can with advice and testing, but I don't know what framework this is built on nor what resources you're using to develop this nor create the libraries / GUI's for the libraries.

The reason behind this request is that, other than a company I now work for, supporting their system and development, other organ software such as Milan Digital Audio's Hauptwerk, has never been and to this day, is not accessible to blind organists, there is some workaround for windows users but that is messy and unstable.. I've made some efforts over the years to try and persuade the developers behind Hauptwerk to find solutions to make their software VoiceOver friendly, but it's either been ignored or promises offered and broken, really sad when someone does this and fails on it.

GrandOrgue is a great system, if we could make it so that VoiceOver can work with the main features such as the main console view and other areas, then we can make this an available and supported option for blind and sight impaired organists. If any of the developers behind this project would like to get in touch with me about this, I'd really appreciate hearing from you.

The last time I tried using this was a couple of years back and it's evident it's moved on so much and seems better to set up than it did. keep up the amazing work. If we could make this app work so that VoiceOver can speak the contents of the console window, interact with buttons, etc, edit functions, etc, then this would be a great solution and you'd have me to support this with video tutorials and demonstrations if that is of interest.

All the best.

Lew

**Comment by oleg68** [comment]
GrandOrgue uses wxWidgets GUI framework https://www.wxwidgets.org/

**Reply by lewisalexander2020** [reply]
I'm not familiar with it, so  please forgive me if my request is a bit of a challenge.

So, here's what happens with GrandOrgue with VoiceOver... 

On initial startup, I'm greeted by midi setup tools which I can happily navigate because it's part of the native mac window and object resource, so voiceover manages it fine, other than announcing "Titel Window"

It loads the demo organ library...

This is where Grandorgue becomes unusable. The main window cannot be used with VoiceOver at all, the only elements found are the Window size, minimize and close, Window title and that's it. there's no page contents, nothing to navigate, making it unusable.

Next test... Panel menus.

opening each window, same as above and no contents detected, nothing to navigate or interact with.

Is there any way we could find a better GUI solution which makes grandorgue accessible with VoiceOver and other screen readers? I'd give anything for this to work.

lew

**Comment by oleg68** [comment]
GrandOrgue does not use Mac native windows for the panels and draws their elements itself.

I've never hear about VoiceOver and don't know what should I do for proper communicating with it.

**Comment by lewisalexander2020** [comment]
Hi, thanks for this, much appreciated.

the overall GUI’s relying on Mac OS’s Window Classes, etc for elements such as setup menus, etc, that’s what I was referring to, but from there, everything else is impossible to use.

Do you have access to a mac? if you do, VoiceOver is built in, not like third party screen readers which cost stupid money and don’t offer the same degree of support, especially in the music industry. My background is as a former developer who worked directly for Apple, developing and servicing VoiceOver for over 13 years until I retired due to medical reasons. My background meant I was working for companies such as Apple, AVID, Native Instruments, in the same speciality, accessibility support and development.

I’ve left virtually all of that behind with exception of a company I’m very loyal to and won’t leave them.

So, VoiceOver is built in, the advantages of this being that if your mac crashes and you’re like me, fully blind, you can completely recover the OS, drive, etc, run diagnostics, etc. it’s a huge game changer, same on every iphone, ipod, ipad, apple TV, apple watch, etc, all designed to give blind users an equal chance in life, removing a lot of the barriers technologies put in place today.

To start VoiceOver, Press the Command and F5 keys together, VoiceOver will start and announce it’s running. initially it’s keyboard based, but I can walk you through setting it up if that would help you understand not only how this works, but blind life if you will.

There’s plenty of online guides to Apple’s VoiceOver for MacOS and some good developer resources which help with making apps accessible.

What is the main developer language you use for GrandOrgue and under what coding software?

The key challenge to getting GrandOrgue to work with VoiceOver is going in to the developer packages and examining how the windows are constructed, then working out how to make the window visible and it’s contents interactive with the VoiceOver cursor and keyboard navigation, etc.

The company I work for, developing their organ software with accessibility in mind, means the barrier is at last being broken, but GrandOrgue could benefit hugely from being VoiceOver friendly.

If I understand this rightly, you use a console method similar to that of Hauptwerk, where the GUI for each console is image based and changes per console, so stop names and positions, quantity of stops, couplers, trems, general and divisional pistons change depending on the design of the organ library.

if this is the case, Could the GUI be adapted in such a way that buttons can be recognised and navigated with the VoiceOver cursor? now this works by either the keyboard mode or having the voiceover cursor coupled to the mouse cursor, so that if you’re using say, a trackpad as an example, what’s under the pointer is spoken. I use this method in diagnostic modes where a window doesn’t have the ability to be navigated with the keyboard’s voiceover interaction due to insufficient mappings.

If I did a quick screen recording to show you, would that help?

lew

**Reply by oleg68** [reply]
> Do you have access to a mac?

Unfortunally, I don't.

> What is the main developer language you use for GrandOrgue and under what coding software?

C++. You can see the whole sourcecode here. The main code of the panels is https://github.com/GrandOrgue/grandorgue/blob/master/src/grandorgue/gui/GOGUIPanel.cpp

> the GUI for each console is image based and changes per console, so stop names and positions, quantity of stops, couplers, trems, general and divisional pistons change depending on the design of the organ library.

Yes it is, exactly.

**Comment by lewisalexander2020** [comment]
do you feel brave enough to try building a hackintosh? or taking a standard PC / laptop and setting it up as a hackintosh? if so, then from there I can then help you with  accessibility methods and give you some guidance where I can.

If we could make GO accessible for blind users, this would be a huge breakthrough and I mean that, as the developers behind Hauptwerk are not interested, I’ve tried for years and surrendered.

All the best.

Lew

**Reply by MStraeten** [reply]
that’s not so trivial as it sounds. The whole application isn’t developed having a mac platform in scope. If you want grandorgue to better integrate into the macOS world then first you need to find developers that are capable to do the job and will spend a lot of spare time in such project. Requesting developers to use hackintosh configurations to meet a requirement they don’t have, isn’t the way foss works…

**Reply by oleg68** [reply]
> do you feel brave enough to try building a hackintosh? or taking a standard PC / laptop and setting it up as a hackintosh?

It contradicts the MacOS license agreement, so I won't do that.

**Comment by lewisalexander2020** [comment]
Sorry if the comment offended in any way. 

All I am trying to do, as an assistive technologies and accessibility advocate / developer within the music industry is not only spread and raise awareness of how music software and instruments would benefit accessibility support for those with sight impairment / sight loss, but also as an organist trying to bring organ software solutions in to the domain of blind and sight impaired organists who rely heavily on assistive technologies to make their lives easier / equal to that of sighted people.

If there’s anyone within the developer pool for this project who is a mac owner / responsible for the mac port of this, then I could discuss with them how we could find a way of getting GO VoiceOver friendly with the windows affected. Whether it’s opening particular resources, changing particular parameters or finding a better GUI method which allows objects, windows and window classes / object classes to be recognised as active elements where a screen reader can navigate and interact with them.

I’ve been in the developer world and left it for health reasons, having been an engineer and developer behind VoiceOver for the Mac and iOS platform, one of over 100 worldwide devs assigned to the project since it’s birth and evolving state. Today, VoiceOver plays a huge part in the music industry for blind people as there’s nothing viable on the windows market either without expensive third party software solutions and development, so today, we have Protools and logic pro which are fully blind friendly, other resources on the market within the music industry which I’ve played a part in, so at least when I die, there’s a legacy.

What I was trying to suggest to you was outside of the standards we have here, purely for you to research and understand how VoiceOver, like other screen readers work, how your developer resources and tools have particular functions available to enable screen readers to integrate effectively with your software and how to get the best out of accessibility development. For you to try VoiceOver on a mac, either on a pure mac or hackintosh, was not an official endorsement by me, just a way to try and find a way of showing you  what this software does, how GO could be made accessible, etc.

If it’s something you can’t do or are not interested in doing, I understand and will leave it at that, but it would be a great shame. If you can demonstrate GO working with a screen reader / develop it so it’s main GUI can be interacted with by VoiceOver and other screen readers, you can then be a huge part in influencing the likes of Milan Digital Audio in accessibility development, if not that, then at least you have a solution that organists like me can use and be proud to use, spread the word about it and build on this further.

lew

**Reply by MStraeten** [reply]
have a look at https://github.com/GrandOrgue/grandorgue/graphs/contributors to get an impression on the developer pool ;)

If you want grandorgue to be voiceover ready then the best way is to get the source, analyze the parts that needs to be adapted, and then contribute ...

**Reply by rousseldenis** [reply]
@MStraeten Eventually, this has been heard : https://github.com/GrandOrgue/grandorgue/pull/1933
